File: app01_java_nft/backend/src/main/java/com/devpowers/nftcerts/api/IdentityController.java
package com.devpowers.nftcerts.api;

import com.devpowers.nftcerts.error.MalformedIdentityException;
import com.devpowers.nftcerts.identity.ArtistIdentity;
import com.devpowers.nftcerts.identity.KycVerificationService;
import jakarta.validation.Valid;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RestController;

import java.util.regex.Pattern;

/** POST /api/identity/verify — artist identity verification (DID / KYC) flow. */
@RestController
public class IdentityController {

    private static final Pattern WALLET_ADDRESS_PATTERN = Pattern.compile("^0x[a-fA-F0-9]{40}$");
    private static final Pattern DID_PATTERN = Pattern.compile("^did:[a-zA-Z0-9]+:.+$");

    private final KycVerificationService kycVerificationService;

    public IdentityController(KycVerificationService kycVerificationService) {
        this.kycVerificationService = kycVerificationService;
    }

    @PostMapping("/api/identity/verify")
    public ResponseEntity<IdentityVerifyResponse> verify(@Valid @RequestBody IdentityVerifyRequest request) {
        if (!WALLET_ADDRESS_PATTERN.matcher(request.walletAddress()).matches()) {
            throw new MalformedIdentityException("Malformed wallet address: " + request.walletAddress());
        }
        if (!DID_PATTERN.matcher(request.did()).matches()) {
            throw new MalformedIdentityException("Malformed DID: " + request.did());
        }

        ArtistIdentity identity = kycVerificationService.verify(request.walletAddress(), request.did(), request.email());

        return ResponseEntity.ok(new IdentityVerifyResponse(identity.isVerified(), identity.getDid(), identity.getWalletAddress()));
    }
}

File: app02_cpp_nft/backend/src/api/IdentityController.cpp
#include "api/IdentityController.h"

#include "error/Exceptions.h"

#include <drogon/HttpAppFramework.h>

#include <regex>

namespace nftcerts::api {

namespace {
const std::regex kWalletAddressPattern("^0x[a-fA-F0-9]{40}$");
const std::regex kDidPattern("^did:[a-zA-Z0-9]+:.+$");
}  // namespace

void registerIdentityRoutes(identity::KycVerificationService& kycVerificationService) {
    drogon::app().registerHandler(
        "/api/identity/verify",
        [&kycVerificationService](const drogon::HttpRequestPtr& req,
                                   std::function<void(const drogon::HttpResponsePtr&)>&& callback) {
            auto body = req->getJsonObject();
            if (!body || !body->isMember("walletAddress") || !body->isMember("did")) {
                throw error::MalformedIdentityException("walletAddress and did are required");
            }
            std::string walletAddress = (*body)["walletAddress"].asString();
            std::string did = (*body)["did"].asString();
            std::string email = body->get("email", "").asString();

            if (!std::regex_match(walletAddress, kWalletAddressPattern)) {
                throw error::MalformedIdentityException("Malformed wallet address: " + walletAddress);
            }
            if (!std::regex_match(did, kDidPattern)) {
                throw error::MalformedIdentityException("Malformed DID: " + did);
            }

            identity::ArtistIdentity identityRecord = kycVerificationService.verify(walletAddress, did, email);

            Json::Value response;
            response["verified"] = identityRecord.verified;
            response["did"] = identityRecord.did;
            response["walletAddress"] = identityRecord.walletAddress;

            callback(drogon::HttpResponse::newHttpJsonResponse(response));
        },
        {drogon::Post});
}

}  // namespace nftcerts::api
